### PR TITLE
FGLM v1.1.0: moved core algorithms to the engine

### PR DESCRIPTION
**Pull request by mahrud**
This PR moves the Groebner basis computation algorithms in FGLM.m2 to the engine, providing significant time improvements over both the best `gb` strategy as well as the previously top-level code in FGLM.m2. For instance, computing a Groebner basis in lex order for the cyclic(7) ideal takes only 32 seconds now, compared to about 6min with the previous version and about 22min with `gb`.

The next update--for M2 v1.16--will move this to become a part of the `gb` command. Also, this source code provides various examples for profiling engine and top-level code which I will put together as a separate documentation node at that point.

### CODE REVIEW

**Comment by DanGrayson** [comment]
This crashes for me under Mac OS X 10.14.  I'll compile in debug mode and find out where.  Do you have access to a Mac?

**Comment by mahrud** [comment]
Does compilation crash or does it crash on tests?
I can find a Mac but if the problem is with the engine code then it might take a while to figure out where the issue is.

**Comment by DanGrayson** [comment]
It fails in running examples:
```
i1 : R1 = QQ[x,y,z];

i2 : I1 = ideal(x^2 + 2*y^2 - y - 2*z, x^2 - 8*y^2 + 10*z - 1, x^2 - 7*y*z);

o2 : Ideal of R1

i3 : R2 = QQ[x,y,z, MonomialOrder => Lex];

i4 : fglm(I1, R2)
-- SIGSEGV
-- stack trace, pid 6122:
level 0 -- return addr: 0x0x10f8f8cf1 -- frame: 0x0x70000170cc60
level 1 -- return addr: 0x0x7fff6cf8fb5d -- frame: 0x0x70000170cc80
level 2 -- return addr: 0x0x114840ab0 -- frame: 0x0x70000170cc90
```

But after compiling in debug mode, I can't reproduce the bug.

**Comment by DanGrayson** [comment]
Well, here's at least something:
```
lldb /Users/dan/src/M2/M2/M2/BUILD/dan/builds.tmp/einsteinium-master/usr-dist/x86_64-Darwin-MacOS-10.14.6/bin/M2-binary
(lldb) target create "/Users/dan/src/M2/M2/M2/BUILD/dan/builds.tmp/einsteinium-master/usr-dist/x86_64-Darwin-MacOS-10.14.6/bin/M2-binary"
Current executable set to '/Users/dan/src/M2/M2/M2/BUILD/dan/builds.tmp/einsteinium-master/usr-dist/x86_64-Darwin-MacOS-10.14.6/bin/M2-binary' (x86_64).
(lldb) run
run
Process 12146 launched: '/Users/dan/src/M2/M2/M2/BUILD/dan/builds.tmp/einsteinium-master/usr-dist/x86_64-Darwin-MacOS-10.14.6/bin/M2-binary' (x86_64)
Macaulay2, version 1.14.0.1
--loading configuration for package "FourTiTwo" from file /Users/dan/Library/Application Support/Macaulay2/init-FourTiTwo.m2
--loading configuration for package "Topcom" from file /Users/dan/Library/Application Support/Macaulay2/init-Topcom.m2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : input "~/src/M2/M2/M2/BUILD/dan/a.m2"
input "~/src/M2/M2/M2/BUILD/dan/a.m2"

ii2 : needsPackage "FGLM"

oo2 = FGLM

oo2 : Package

ii3 : R1 = QQ[x,y,z];

ii4 : I1 = ideal(x^2 + 2*y^2 - y - 2*z, x^2 - 8*y^2 + 10*z - 1, x^2 - 7*y*z);

oo4 : Ideal of R1

ii5 : R2 = QQ[x,y,z, MonomialOrder => Lex];

ii6 : fglm(I1, R2)
Process 12146 stopped
* thread #2, stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00007fff6a0233d8 libc++abi.dylib`__dynamic_cast + 26
libc++abi.dylib`__dynamic_cast:
->  0x7fff6a0233d8 <+26>: movq   -0x10(%rax), %r13
    0x7fff6a0233dc <+30>: addq   %rdi, %r13
    0x7fff6a0233df <+33>: movq   -0x8(%rax), %r14
    0x7fff6a0233e3 <+37>: movq   %rdx, -0x88(%rbp)
Target 0: (M2-binary) stopped.
(lldb) bt
bt
* thread #2, stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00007fff6a0233d8 libc++abi.dylib`__dynamic_cast + 26
    frame #1: 0x0000000100292064 M2-binary`MutableMat<DMat<M2::ARingQQGMP> >::addMultipleTo(MutableMatrix const*, MutableMatrix const*) [inlined] MutableMat<DMat<M2::ARingQQGMP> > const* MutableMatrix::cast_to_MutableMat<DMat<M2::ARingQQGMP> >(this=0x0000000100072bdf) const at mat.hpp:92:12 [opt]
    frame #2: 0x000000010029203e M2-binary`MutableMat<DMat<M2::ARingQQGMP> >::addMultipleTo(MutableMatrix const*, MutableMatrix const*) [inlined] DMat<M2::ARingQQGMP> const* MutableMatrix::coerce_const<DMat<M2::ARingQQGMP> >(this=0x0000000100072bdf) const at mat.hpp:106 [opt]
    frame #3: 0x000000010029203e M2-binary`MutableMat<DMat<M2::ARingQQGMP> >::addMultipleTo(this=0x000000010553f100, A=<unavailable>, B=0x0000000100072bdf) at mutablemat-imp.hpp:140 [opt]
    frame #4: 0x000000010002c83f M2-binary`::engine_star__7(M=<unavailable>, N=<unavailable>) at engine.dd:202:22 [opt]
    frame #5: 0x0000000100072cff M2-binary`actors_star_(lhs=<unavailable>, rhs=0x000000010444ca40) at actors.d:313:17 [opt]
    ...
(lldb) up
up
M2-binary was compiled with optimization - stepping may behave oddly; variables may not be available.
frame #1: 0x0000000100292064 M2-binary`MutableMat<DMat<M2::ARingQQGMP> >::addMultipleTo(MutableMatrix const*, MutableMatrix const*) [inlined] MutableMat<DMat<M2::ARingQQGMP> > const* MutableMatrix::cast_to_MutableMat<DMat<M2::ARingQQGMP> >(this=0x0000000100072bdf) const at mat.hpp:92:12 [opt]
   89  	  template <typename MatType>
   90  	  const MutableMat<MatType> *cast_to_MutableMat() const
   91  	  {
-> 92  	    return dynamic_cast<const MutableMat<MatType> *>(this);
    	           ^
   93  	  }
   94  	
   95  	  template <typename MatT>
(lldb) p this
p this
(MutableMatrix *) $0 = 0x0000000100072bdf
(lldb) p *this
p *this
(MutableMatrix) $1 = {
  MutableEngineObject = (mHashValue = 2210445071)
}
```

**Comment by DanGrayson** [comment]
I think we've missed the 1.15 release on this one.

**Comment by mahrud** [comment]
I still haven't been able to reproduce this, by the way.

**Comment by DanGrayson** [comment]
Let's reactivate this and I'll test it before 1.16.

**Comment by mahrud** [comment]
Thanks @DanGrayson. Do you still get that build error when you compile? At least on the CMake branch it compiles without a hitch.

**Comment by DanGrayson** [comment]
We'll see -- I haven't tried it yet recently.

File: M2/Macaulay2/e/engine.h [hunk 1]
```diff
@@ -1473,6 +1473,27 @@ extern "C" {
      A encodes both the L and the U part, as in lapack.
    */
 
+  /**
+     Hi, this is rawLUincremental.
```
Comment (mikestillman) [inline M2/Macaulay2/e/engine.h:1477]: Hi, just testing github comments and replies

**Comment by mikestillman** [comment]
Testing  a link https://github.com/Macaulay2/M2/pull/1047#issuecomment-633561179

**Pull request merged**